File: algorithmic-toolbox/W2/06 Last Digit of the Sum of Fibonacci Numbers/fibonacci_sum_last_digit.py
# Uses python3
import sys

def get_fibonacci_last_digit(n):
    if (n < 1):
        return n

    prev = 0
    curr = 1

    for _ in range(n - 1):
        prev, curr = curr % 10, (prev + curr) % 10

    return curr % 10

def get_fibonacci_sum(n):
    # Little trick:
    # Pisano period modulo 10 is 60
    # Sum of n Fibonacci numbers is F(n + 2) - 1
    last_digit = get_fibonacci_last_digit((n + 2) % 60)
    sum_last_digit = (last_digit - 1) if (last_digit != 0) else 9
    return sum_last_digit

if __name__ == '__main__':
    input = sys.stdin.read()
    n = int(input)
    print(get_fibonacci_sum(n))

File: algorithmic-toolbox/W2/06 Last Digit of the Sum of Fibonacci Numbers/FibonacciSumLastDigit.java
import java.util.*;

public class FibonacciSumLastDigit {
    private static final int PISANO_PERIOD_MOD10 = 60;

    private static int getFibonacciLastDigit(long n) {
        if (n < 1)
            return (int) n;

        int prev = 0;
        int curr = 1;
        for (long i = 2; i <= n; i++) {
            int next = (prev + curr) % 10;
            prev = curr;
            curr = next;
        }
        return curr % 10;
    }

    private static int getFibonacciSum(long n) {
        // Little trick:
        // Pisano period modulo 10 is 60
        // Sum of n Fibonacci numbers is F(n + 2) - 1
        int lastDigit = getFibonacciLastDigit((n + 2) % PISANO_PERIOD_MOD10);
        int sumLastDigit = (lastDigit == 0) ? 9 : (lastDigit - 1);
        return sumLastDigit;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        long s = getFibonacciSum(n);
        System.out.println(s);
    }
}